src/tests.rs

```diff
@@ -199,6 +206,85 @@ fn prop_send_recv_half_closed() {
     QuickCheck::new().tests(7).quickcheck(prop as fn(_))
 }
 
+/// This test simulates two endpoints of a Yamux connection which may be unable to
+/// write simultaneously but can make progress by reading. If both endpoints
+/// don't read in-between trying to finish their writes, a deadlock occurs.
+#[test]
+#[ignore]
+fn write_deadlock() {
+    let _ = env_logger::try_init();
+    let mut pool = LocalPool::new();
+
+    // We make the message to transmit large enough s.t. the "server"
+    // is forced to start writing (i.e. echoing) the bytes before
+    // having read the entire payload.
+    let msg = vec![1u8; 1024 * 1024];
+
+    // Create a bounded channel representing the underlying "connection".
+    // Each endpoint gets a name and a bounded capacity for its outbound
+    // channel (which is the other's inbound channel).
+    let (server_endpoint, client_endpoint) = bounded::channel(("S", 1024), ("C", 1024));
+
+    // Create and spawn a "server" that echoes every message back to the client.
+    let server = Connection::new(server_endpoint, Config::default(), Mode::Server);
+    pool.spawner().spawn_obj(async move {
+        crate::into_stream(server).try_for_each_concurrent(
+            None, |mut stream| async move {
+                {
+                    let (mut r, mut w) = AsyncReadExt::split(&mut stream);
+                    // Write back the bytes received. This may buffer internally.
+                    futures::io::copy(&mut r, &mut w).await?;
+                }
+                log::debug!("S: stream {} done.", stream.id());
+                stream.close().await?;
+                Ok(())
+            })
+            .await
+            .expect("server failed")
+    }.boxed().into()).unwrap();
+
+    // Create and spawn a "client" that sends messages expected to be echoed
+    // by the server.
+    let client = Connection::new(client_endpoint, Config::default(), Mode::Client);
+    let mut ctrl = client.control();
+
+    // Continuously advance the Yamux connection of the client in a background task.
+    pool.spawner().spawn_obj(
+        crate::into_stream(client).for_each(|_| {
+            panic!("Unexpected inbound stream for client");
+            #[allow(unreachable_code)]
+            future::ready(())
+        }).boxed().into()
+    ).unwrap();
+
+    // Handles to tasks on which to wait for completion.
+    let mut handles = Vec::new();
```

Comment:
Given that there is only ever a single handler, is the `Vec` and the `for` loop below necessary?

Reply:
That is a left-over :+1: .

Reply:
Addressed in https://github.com/paritytech/yamux/pull/104/commits/2ed415012803c22d65d539bc5c7417cbda9dfd91.